http://bugs.gentoo.org/246266

From 5136519230209e4362e477c38d92f7d25203255f Mon Sep 17 00:00:00 2001
From: Joel E. Denny <jdenny@ces.clemson.edu>
Date: Fri, 7 Nov 2008 17:20:44 -0500
Subject: [PATCH] Clean up %skeleton and %language priority implementation.

* src/getargs.c (skeleton_prio): Use default_prio rather than 2, and
remove static qualifier because others will soon need to see it.
(language_prio): Likewise.
(getargs): Use command_line_prio rather than 0.
* src/getargs.h (command_line_prio, grammar_prio, default_prio): New
enum fields.
(skeleton_prio): Extern it.
(language_prio): Extern it.
* src/parse-gram.y: Use grammar_prio rather than 1.
---
 ChangeLog        |   13 +++
 src/getargs.c    |   10 ++-
 src/getargs.h    |    3 +
 src/parse-gram.c |  226 +++++++++++++++++++++++++++---------------------------
 src/parse-gram.h |    8 +-
 src/parse-gram.y |    4 +-
 6 files changed, 141 insertions(+), 123 deletions(-)

diff --git a/src/getargs.c b/src/getargs.c
index c36cabf..a8d1c54 100644
--- a/src/getargs.c
+++ b/src/getargs.c
@@ -72,9 +72,9 @@ static struct bison_language const valid_languages[] = {
   { "", "", "", "", false }
 };
 
-static int skeleton_prio = 2;
+int skeleton_prio = default_prio;
 const char *skeleton = NULL;
-static int language_prio = 2;
+int language_prio = default_prio;
 struct bison_language const *language = &valid_languages[0];
 const char *include = NULL;
 
@@ -540,11 +540,11 @@ getargs (int argc, char *argv[])
 	break;
 
       case 'L':
-	language_argmatch (optarg, 0, NULL);
+	language_argmatch (optarg, command_line_prio, NULL);
 	break;
 
       case 'S':
-	skeleton_arg (AS_FILE_NAME (optarg), 0, NULL);
+	skeleton_arg (AS_FILE_NAME (optarg), command_line_prio, NULL);
 	break;
 
       case 'T':
diff --git a/src/getargs.h b/src/getargs.h
index c9e3f48..8d27e71 100644
--- a/src/getargs.h
+++ b/src/getargs.h
@@ -24,11 +24,13 @@
 #include "location.h"
 
 extern char *program_name;
+enum { command_line_prio, grammar_prio, default_prio };
 
 /* flags set by % directives */
 
 /* for -S */
 extern char const *skeleton;
+extern int skeleton_prio;
 
 /* for -I */
 extern char const *include;
@@ -68,6 +70,7 @@ struct bison_language
   bool add_tab;
 };
 
+extern int language_prio;
 extern struct bison_language const *language;
 
 /*-----------.
diff --git a/src/parse-gram.c b/src/parse-gram.c
index 76d25ef..ecb1cae 100644
--- a/src/parse-gram.c
+++ b/src/parse-gram.c
@@ -1992,11 +1992,11 @@
 
   case 18:
 
 /* Line 1455 of yacc.c  */
 #line 257 "parse-gram.y"
-    { language_argmatch ((yyvsp[(2) - (2)].chars), 1, &(yylsp[(1) - (2)])); }
+    { language_argmatch ((yyvsp[(2) - (2)].chars), grammar_prio, &(yylsp[(1) - (2)])); }
     break;
 
   case 19:
 
 /* Line 1455 of yacc.c  */
@@ -2108,11 +2108,11 @@
             }
           strcpy (skeleton_build + dir_length, skeleton_user);
           skeleton_user = uniqstr_new (skeleton_build);
           free (skeleton_build);
         }
-      skeleton_arg (skeleton_user, 1, &(yylsp[(1) - (2)]));
+      skeleton_arg (skeleton_user, grammar_prio, &(yylsp[(1) - (2)]));
     }
     break;
 
   case 31:
 

From 58bd33b7fc5d13fbd4b61404e1e878ab3b3c16da Mon Sep 17 00:00:00 2001
From: Joel E. Denny <jdenny@ces.clemson.edu>
Date: Tue, 4 Nov 2008 15:03:00 -0500
Subject: [PATCH] Fix user actions without a trailing semicolon.

Reported by Sergei Steshenko at
<http://lists.gnu.org/archive/html/bug-bison/2008-11/msg00001.html>.
* THANKS (Sergei Steshenko): Add.
* src/scan-code.l (SC_RULE_ACTION): Fix it.
* tests/regression.at (Fix user actions without a trailing semicolon):
New test case.
---
 ChangeLog           |   10 ++++++++++
 THANKS              |    1 +
 src/scan-code.l     |    4 ++--
 tests/regression.at |   26 +++++++++++++++++++++++++-
 4 files changed, 38 insertions(+), 3 deletions(-)

diff --git a/src/scan-code.l b/src/scan-code.l
index 630d45d..71c9076 100644
--- a/src/scan-code.l
+++ b/src/scan-code.l
@@ -170,7 +170,7 @@ splice	 (\\[ \f\t\v]*\n)*
 
   "{"  STRING_GROW; ++braces_level;
   "}"  {
-    bool outer_brace = --braces_level < 0;
+    bool outer_brace = --braces_level == 0;
 
     /* As an undocumented Bison extension, append `;' before the last
        brace in braced code, so that the user code can omit trailing

From d9a9b96bb51bf5430deaba9c5f274d5f56f0fea9 Mon Sep 17 00:00:00 2001
From: Joel E. Denny <jdenny@ces.clemson.edu>
Date: Fri, 7 Nov 2008 17:21:23 -0500
Subject: [PATCH] Don't add a semicolon to actions for %skeleton or %language.

It breaks Java test cases as reported by Akim Demaille.
* src/scan-code.l: Implement.
---
 ChangeLog       |    6 ++++++
 src/scan-code.l |    7 +++++--
 2 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/src/scan-code.l b/src/scan-code.l
index 71c9076..13a78c2 100644
--- a/src/scan-code.l
+++ b/src/scan-code.l
@@ -175,7 +175,9 @@ splice	 (\\[ \f\t\v]*\n)*
     /* As an undocumented Bison extension, append `;' before the last
        brace in braced code, so that the user code can omit trailing
        `;'.  But do not append `;' if emulating Yacc, since Yacc does
-       not append one.
+       not append one.  Also, some output languages (like Java) do not
+       accept an extra semicolon, so don't append if the user specified
+       a skeleton or language.
 
        FIXME: Bison should warn if a semicolon seems to be necessary
        here, and should omit the semicolon if it seems unnecessary
@@ -186,7 +188,8 @@ splice	 (\\[ \f\t\v]*\n)*
        should also diagnose other Bison extensions like %yacc.
        Perhaps there should also be a GCC-style --pedantic-errors
        option, so that such warnings are diagnosed as errors.  */
-    if (outer_brace && ! yacc_flag)
+    if (outer_brace && !yacc_flag && language_prio == default_prio
+        && skeleton_prio == default_prio)
       obstack_1grow (&obstack_for_string, ';');
 
     STRING_GROW;
